src/GeniusYield/Types/PlutusVersion.hs

```diff
@@ -24,57 +26,74 @@ import           GHC.TypeLits
+    geq _ _                       = Nothing
 
 instance GCompare SingPlutusVersion where
     gcompare SingPlutusV1 SingPlutusV1 = GEQ
-    gcompare SingPlutusV1 SingPlutusV2 = GLT
+    gcompare SingPlutusV1 _            = GLT
     gcompare SingPlutusV2 SingPlutusV1 = GGT
     gcompare SingPlutusV2 SingPlutusV2 = GEQ
+    gcompare SingPlutusV2 _            = GLT
+    gcompare SingPlutusV3 SingPlutusV1 = GGT
+    gcompare SingPlutusV3 SingPlutusV2 = GGT
+    gcompare SingPlutusV3 SingPlutusV3 = GEQ
 
 type family PlutusVersionToApi (v :: PlutusVersion) :: Type where
     PlutusVersionToApi 'PlutusV1 = Api.PlutusScriptV1
     PlutusVersionToApi 'PlutusV2 = Api.PlutusScriptV2
+    PlutusVersionToApi 'PlutusV3 = Api.PlutusScriptV3
 
 singPlutusVersionToApi :: SingPlutusVersion v -> Api.S.PlutusScriptVersion (PlutusVersionToApi v)
 singPlutusVersionToApi SingPlutusV1 = Api.PlutusScriptV1
 singPlutusVersionToApi SingPlutusV2 = Api.PlutusScriptV2
+singPlutusVersionToApi SingPlutusV3 = Api.PlutusScriptV3
 
 fromSingPlutusVersion :: SingPlutusVersion v -> PlutusVersion
+fromSingPlutusVersion SingPlutusV3 = PlutusV3
 fromSingPlutusVersion SingPlutusV2 = PlutusV2
 fromSingPlutusVersion SingPlutusV1 = PlutusV1
 
+-- | Type family to compare `PlutusVersion`s
+type family CmpPlutusVersion (v :: PlutusVersion) (u :: PlutusVersion) :: Ordering where
+  CmpPlutusVersion 'PlutusV1 'PlutusV1 = 'EQ
+  CmpPlutusVersion 'PlutusV1 _ = 'LT
+  CmpPlutusVersion 'PlutusV2 'PlutusV1 = 'GT
+  CmpPlutusVersion 'PlutusV2 'PlutusV2 = 'EQ
+  CmpPlutusVersion 'PlutusV2 _ = 'LT
+  CmpPlutusVersion 'PlutusV3 'PlutusV1 = 'GT
+  CmpPlutusVersion 'PlutusV3 'PlutusV2 = 'GT
+  CmpPlutusVersion 'PlutusV3 'PlutusV3 = 'EQ
+
 -- | Constraint that @v >= u@.
 --
--- Used to allow using V2 transaction features only in transactions with V2 inputs.
+-- If transaction is making use of V2 features (such as reference inputs) then as these cannot be represented in script context of V1 scripts, we need to ensure that the involved script version is at least V2. Likewise for other versions.
 class VersionIsGreaterOrEqual (v :: PlutusVersion) (u :: PlutusVersion)
+instance (If ((v `CmpPlutusVersion` u) == 'LT) (TypeError ('Text "Given version " ':<>: ShowType v ':<>: 'Text ", is not greater or equal to " ':<>: ShowType u)) (() :: Constraint)) => VersionIsGreaterOrEqual v u
 
--- | Any version is greater or equal to 'PlutusV1'
-instance VersionIsGreaterOrEqual 'PlutusV1 'PlutusV1
-
--- | Any version is greater or equal to 'PlutusV1'
-instance VersionIsGreaterOrEqual 'PlutusV2 'PlutusV1
-
--- | Only 'PlutusV2' is greater or equal to itself at the moment.
-instance VersionIsGreaterOrEqual 'PlutusV2 'PlutusV2
-
--- | Explicitly ruled out instance.
-instance TypeError ('Text "V1 is not >= V2") => VersionIsGreaterOrEqual 'PlutusV1 'PlutusV2
+-- | Constraint that @v > u@.
+class VersionIsGreater (v :: PlutusVersion) (u :: PlutusVersion)
+instance ((v `CmpPlutusVersion` u) ~ 'GT) => VersionIsGreater v u
```

Comment:
Not sure if this extra addition of `CmpPlutusVersion` is necessary. It likely worsens type errors (and type inferences) and doesn't seem to gain anything? Unless I missed something.

Reply:
It does indeed worsens type errors and inferences... I initially drafted `CmpPlutusVersion` to make the instances less verbose. Let me revisit it.

Reply:
I also wanted to avoid code duplication b/w `VersionIsGreaterOrEqual` and `VersionIsGreater`. But taking a look on improving it.

Reply:
I have made updates in [this](https://github.com/geniusyield/atlas/pull/327/commits/6f1f770ea3ffa7c1b8cd6db21262ffd43a619053) commit. Type errors/inference must be improved now as this change led to spotting of some errors in the code.

Reply:
Does this mean that users _must_ put type annotations everywhere now? That's not good...

Reply:
Nope! Those `PlutusV2` annotations were added because one could put `PlutusV3` as well and so compiler is not sure.